cpp_test_suite/new_tests/cxx_pipe_event.cpp

```diff
@@ -0,0 +1,315 @@
+        try
+        {
+            device1 = new DeviceProxy(device1_name);
+            device1->ping();
+        }
+        catch (CORBA::Exception &e)
+        {
+            Except::print_exception(e);
+            exit(-1);
+        }
+    }
+
+    virtual ~SUITE_NAME()
+    {
+        delete device1;
+    }
+
+    static SUITE_NAME *createSuite()
+    {
+        return new SUITE_NAME();
+    }
+
+    static void destroySuite(SUITE_NAME *suite)
+    {
+        delete suite;
+    }
+
+//
+// Tests -------------------------------------------------------
+//
+
+// Test Tango device server signal features
+
+    void test_pipe_event__loop(void)
+    {
+        cb_executed = 0;
+        cb_err = 0;
+
+        class EventCallBack : public Tango::CallBack
+        {
+        public:
+            EventCallBack(PipeEventTestSuite__loop* outer):
+            root_blob_name(""),
+            nb_data(0),
+            outer(outer)
+            {}
+            std::string root_blob_name;
+            size_t	nb_data;
+            PipeEventTestSuite__loop* outer;
+            void push_event(Tango::PipeEventData* event_data){
+                {
+                    std::lock_guard<std::mutex> lk(outer->m);
+                    outer->cb_executed++;
+
+
+                    try
+                    {
+                        coutv << "EventCallBack::push_event(): called pipe " << event_data->pipe_name << " event "
+                              << event_data->event << "\n";
+                        if (!event_data->err)
+                        {
+                            coutv << "Received pipe event for pipe " << event_data->pipe_name << std::endl;
+//			coutv << *(event_data->pipe_value) << std::endl;
+                            root_blob_name = event_data->pipe_value->get_root_blob_name();
+
+                            if (root_blob_name == "PipeEventCase4")
+                            {
+                                std::vector<Tango::DevLong> v_dl;
+                                (*(event_data->pipe_value))["Martes"] >> v_dl;
+                                nb_data = v_dl.size();
+                            }
+                        }
+                        else
+                        {
+                            coutv << "Error sent to callback" << std::endl;
+//			Tango::Except::print_error_stack(event_data->errors);
+                        }
+                }
+                catch (...)
+                {
+                    coutv << "EventCallBack::push_event(): could not extract data !\n";
+                }
+            }
+
+            outer->cv.notify_all();
+        }
+    };
+
+        EventCallBack cb(this);
+
+//
+// subscribe to a pipe event
+//
+
+        int eve_id1 = device1->subscribe_event("RWPipe",Tango::PIPE_EVENT,&cb);
+
+//
+// The callback should have been executed once
+//
+        {
+            std::unique_lock<std::mutex> lk(m);
+            cv.wait_for(lk, std::chrono::milliseconds(1000));
+        }
```

Comment:
Can you extract this to a function? Maybe just a lambda on top of the testcase, like:

```cpp
auto wait_for_event = [&lk, &m, &cv](int millis)
{
            std::unique_lock<std::mutex> lk(m);
            cv.wait_for(lk, std::chrono::milliseconds(1000));
}
```

And then just call it instead of copy-pasting the whole block.